### PR TITLE
New option to store unsafe classes on shared cache

### PR DESCRIPTION
**Pull request by doomerxe**
Allow unsafe classes to be stored and found in the shared class cache by default and add a new command line option -XX:[+/-]ShareUnsafeClasses

Closes: #4985

Doc issue eclipse/openj9-docs#351

Changes are based on @sogutbera 's work and comments on #6896

Signed-off-by: Jiahan Xi <doomerXe@gmail.com>

### CODE REVIEW

**Comment by doomerxe** [comment]
Hi @hangshao0! Could you please review this? Thank you so much!

File: runtime/bcutil/ROMClassCreationContext.hpp [hunk 1]
```diff
@@ -293,7 +290,38 @@ class ROMClassCreationContext
...
+		if ((NULL != _javaVM)
+		&& (NULL != _javaVM->sharedClassConfig)
+		) {
+			isAnonDefineClassSharingEnabled = J9_ARE_ALL_BITS_SET(_javaVM->sharedClassConfig->runtimeFlags, J9SHR_RUNTIMEFLAG_ENABLE_SHAREANONYMOUSCLASSES);
+			isUnsafeDefineClassSharingEnabled = J9_ARE_ALL_BITS_SET(_javaVM->sharedClassConfig->runtimeFlags, J9SHR_RUNTIMEFLAG_ENABLE_SHAREUNSAFECLASSES);
+		}
+
+		if (isClassAnon()) {
+			/* 
+			 * Assert if ShareAnonymousClasses is disabled
```
Comment (hangshao0) [inline runtime/bcutil/ROMClassCreationContext.hpp]: Better to add explanation why we assert here.

File: runtime/bcutil/ROMClassCreationContext.hpp [hunk 2]
```diff
@@ -293,7 +290,38 @@ class ROMClassCreationContext
...
+		 * 3. -XX:-ShareAnonymousClasses -XX:+ShareUnsafeClasses, this will only share unsafe classes that are not anonymous
+		 * 4. -XX:-ShareAnonymousClasses -XX:-ShareUnsafeClasses, this will share neither.
+		 * The current default behavior is the 1st option.
+		 */
+		bool isEnabled = false;
+
+		bool isAnonDefineClassSharingEnabled = false;
+		bool isUnsafeDefineClassSharingEnabled = false;
+		if ((NULL != _javaVM)
+		&& (NULL != _javaVM->sharedClassConfig)
```
Comment (hangshao0) [inline runtime/bcutil/ROMClassCreationContext.hpp]: I don't think we need to do NULL check on _javaVM and _javaVM->sharedClassConfig here. They cannot be NULL here.

**Comment by hangshao0** [comment]
Why is this closed ?

**Comment by hangshao0** [comment]
@doomerxe  Did you sign ECA ? https://www.eclipse.org/legal/ECA.php

**Comment by doomerxe** [comment]
> @doomerxe Did you sign ECA ? https://www.eclipse.org/legal/ECA.php

Yes, I did that. I passed the check on the first time

File: runtime/bcutil/ROMClassCreationContext.hpp [hunk 3]
```diff
@@ -293,7 +290,33 @@ class ROMClassCreationContext
...
+		 */
+		bool isEnabled = false;
+
+		bool isAnonDefineClassSharingEnabled = J9_ARE_ALL_BITS_SET(_javaVM->sharedClassConfig->runtimeFlags, J9SHR_RUNTIMEFLAG_ENABLE_SHAREANONYMOUSCLASSES);
+		bool isUnsafeDefineClassSharingEnabled = J9_ARE_ALL_BITS_SET(_javaVM->sharedClassConfig->runtimeFlags, J9SHR_RUNTIMEFLAG_ENABLE_SHAREUNSAFECLASSES);
+
+		if (isClassAnon()) {
+			/* 
+			 * Since the class loader shared classes enable flag is set properly and it is checked in the isClassLoaderSharedClassesEnabled() function
+			 * before this function is called in isROMClassShareable(). Thus, we can just assert if ShareAnonymousClasses is disabled
```
Comment (hangshao0) [inline runtime/bcutil/ROMClassCreationContext.hpp]: Please change "assert if ShareAnonymousClasses is disabled" to "assert isAnonDefineClassSharingEnabled is true"

**Comment by hangshao0** [comment]
Please keep the original commit message of https://github.com/eclipse/openj9/pull/6896. And you can squash the two commits into one.

File: runtime/shared/shrclssup.c [hunk 4]
```diff
@@ -145,6 +145,13 @@ IDATA J9VMDllMain(J9JavaVM* vm, IDATA stage, void* reserved)
 					runtimeFlags &= (~J9SHR_RUNTIMEFLAG_ENABLE_SHAREANONYMOUSCLASSES);
 				}
 
+				/* Check for -XX:+ShareUnsafeClasses and -XX:-ShareUnsafeClasses; whichever comes later wins. Enable is set by default so we just need to disable when that's the case. */
+				argIndex1 = FIND_ARG_IN_VMARGS(OPTIONAL_LIST_MATCH, VMOPT_XXENABLESHAREUNSAFECLASSES, NULL);
```
Comment (pshipton) [inline runtime/shared/shrclssup.c]: These checks need to consume the args, see https://github.com/eclipse/openj9/issues/6870#issuecomment-529021381

Comment (doomerxe) [reply runtime/shared/shrclssup.c]: Thank you for the information! I will also change the error message later.

Comment (pshipton) [reply runtime/shared/shrclssup.c]: Pls explain. What error message do you plan to change?

File: runtime/shared/shrclssup.c [hunk 5]
```diff
@@ -130,21 +130,28 @@ IDATA J9VMDllMain(J9JavaVM* vm, IDATA stage, void* reserved)
 				/* Check for -XX:ShareClassesEnableBCI and -XX:ShareClassesDisableBCI; whichever comes later wins.
 				 * These options should be checked before parseArgs() to allow -Xshareclasses:[enable|disable]BCI to override this option.
 				 */
-				argIndex1 = FIND_ARG_IN_VMARGS(OPTIONAL_LIST_MATCH, VMOPT_XXSHARECLASSESENABLEBCI, NULL);
-				argIndex2 = FIND_ARG_IN_VMARGS(OPTIONAL_LIST_MATCH, VMOPT_XXSHARECLASSESDISABLEBCI, NULL);
+				argIndex1 = FIND_AND_CONSUME_ARG(OPTIONAL_LIST_MATCH, VMOPT_XXSHARECLASSESENABLEBCI, NULL);
```
Comment (pshipton) [inline runtime/shared/shrclssup.c]: I don't think any of these options should be using OPTIONAL_LIST_MATCH, they should all use EXACT_MATCH

File: runtime/shared/shrclssup.c (hunk 4 above)
Comment (doomerxe) [reply runtime/shared/shrclssup.c]: According to Jack's comment, I will change the error message of `Command-line option unrecognised:` into some thing like `.... should only be used with -Xshareclasses option` if these options are used without `-Xshareclasses`

Comment (pshipton) [reply runtime/shared/shrclssup.c]: I don't think we should do that. What we can do is quietly consume the options if -Xshareclasses doesn't do it. See https://github.com/eclipse/openj9/blob/dde4b86e24f7e01f8f70c8d16bde45b0cb4efdd5/runtime/vm/jvminit.c#L2541

**Comment by pshipton** [comment]
jenkins test sanity,extended plinux,win,osx jdk8,jdk11

**Comment by pshipton** [comment]
https://ci.eclipse.org/openj9/job/Test_openjdk8_j9_sanity.functional_ppc64le_linux_Personal/243
just stopped running testing between test suites, no messages, no artifacts.

**Comment by pshipton** [comment]
jenkins test sanity plinux jdk8

**Comment by pshipton** [comment]
Assuming the last bit of testing passes, I'm going to merge this as-is. This gets this change in the 0.17 release for which the code split is on the weekend. The remaining issues are pre-existing problems and they can be fixed in another PR, which when ready we can double deliver to also include in the 0.17 release if necessary.

File: runtime/shared/shrclssup.c (hunk 5 above)
Comment (pshipton) [reply runtime/shared/shrclssup.c]: Reminder the OPTIONAL_LIST_MATCH problem should be fixed, in another PR.

**Pull request merged**